tm-tool/textmapper/resources/org/textmapper/tool/templates/go_types.ltp

```diff
@@ -41,7 +41,7 @@ ${if tokens = opts.reportTokens
 ${end-}
 ${call customRules-}
 	}
-	"log".Fatalf("unknown node type %v\n", n.Type())
+	panic("fmt".Errorf("ast: unknown node type %v", n.Type()))
```

Comment:
Yeah, I agree that this is not a fatal error but I also feel uneasy about unconditionally panic-ing here (at my company we prefer not to panic). How about we extract this statement into a separate template and then call it here (`${call noTokenPanic-}`)?
```
${template noTokenPanic-}
	panic("fmt".Errorf("ast: unknown node type %v", n.Type()))
${end}
```

Reply:
`Fatal` is worse than panic'ing, because it gives even less information about why the program is exiting. If you don't want to panic, you should avoid `Fatal` as well.

If panic is not appropriate, because it is not a fatal error, then the idiomatic thing to do in Go is to return the error through an additional return value, or to store the error state somewhere that it can re-appear at an appropriate time.

For example, if you were doing buffered IO, you might not want to check the error all the time:

```go
buf := bufio.NewWriter(os.Stdout)
_, _ = buf.Write(...)
_, _ = buf.Write(...)
_, _ = buf.Write(...)
_, _ = buf.Write(...)
err := buf.Flush()
if err != nil {
}
```

This takes advantage of the fact that errors can be sticky, and that if an error is present, then a `Write` is effectively a no-op.

However, if you had something that was more similar to this:

```go
buf := bufio.NewWriter(os.Stdout)
for i := 0; i < 1000000; i++ {
  _, _ = buf.Write(...)
  if err != nil {
  }
}
```

... then you would want to abort the loop when you hit the first error to avoid wasting too much compute.

I used bufio as an example, but another way of doing error handling would be as so, with the bufio [used for scanning](https://tip.golang.org/pkg/bufio/#Scanner):

```go
scanner := bufio.NewScanner(os.Stdin)
for scanner.Scan() {
    fmt.Println(scanner.Text()) // Println will add back the final '\n'
}
if err := scanner.Err(); err != nil {
    fmt.Fprintln(os.Stderr, "reading standard input:", err)
}
```

Apologies if this is all obvious, just wanted to provide some alternatives for you, since I think that Fatal is particularly not great here.

Reply:
The other comment about `panic` is that really it should be used for unrecoverable circumstances which are programmer errors, rather than 'runtime errors'.

`Fatal` should really be reserved for the `main` function, or at least "the application", and not called by libraries.

Reply:
log.Fatal is very appropriate for broken invariants which should crash the program ASAP (a pattern in server-side development, which typically triggers a release rollback). It is not appropriate here of course.

In this case, returning 'nil' is probably good enough for most users. I'm also fine with having it 'panic' by default, as you suggested. I merely propose to extract it into a template so that users can remove the panic if they want.

Reply:
> In this case, returning 'nil' is probably good enough for most users. I'm also fine with having it 'panic' by default, as you suggested. I merely propose to extract it into a template so that users can remove the panic if they want.

I'd prefer panic over retuning nil. Mainly as that helps locate where the issue is. The returned nil value may otherwise be propagated into other structures and be discovered much later.